File: haskell/binary_tree.hs
data Tree
  = Empty
  | Node Tree Double Tree
  deriving (Show)

add :: Double -> Tree -> Tree
add elt Empty = Node Empty elt Empty
add elt (Node l value r)
  | elt == value = Node l value r
  | elt < value  = Node (add elt l) value r
  | otherwise    = Node l value (add elt r)

count_nodes :: Num a => Tree -> a
count_nodes Empty = 0
count_nodes (Node l value r) = 
  1 + count_nodes l + count_nodes r
  
sum_nodes :: Tree -> Double
sum_nodes Empty = 0.0
sum_nodes (Node l value r) = 
  value + sum_nodes l + sum_nodes r

flatten :: Tree -> [Double]
flatten Empty = []
flatten (Node l value r) = 
  [value] ++ flatten l ++ flatten r

average_value :: Tree -> Double
average_value Empty = 0.0
average_value (Node l value r) = 
  sum / count
  where (sum, count) = get_sum_count (Node l value r)

get_sum_count :: Tree -> (Double, Double)
get_sum_count Empty = (0.0, 0.0)
get_sum_count (Node l value r) = (value + lsum + rsum, 1 + lcount + rcount)
  where (lsum, lcount) = get_sum_count l
        (rsum, rcount) = get_sum_count r

average_value1 :: Tree -> Double
average_value1 Empty = 0.0
average_value1 (Node l value r)
  = sum nodes / fromIntegral(length nodes)
  where nodes = flatten (Node l value r)

average_value2 :: Tree -> Double
average_value2 Empty = 0.0
average_value2 (Node l value r) = sum / count
  where sum = sum_nodes (Node l value r)
        count = count_nodes (Node l value r)

File: prolog/binary_tree.pl
% This file contains an implementation of a binary tree in prolog. 
% The 'binary_tree.hs' in the haskell folder contains an implementation of 
% exactly the same binary tree in haskell. 

% empty
% node empty 5.5 empty

add(Elt, empty, node(empty, Elt, empty)).
add(Elt, node(L, Value, R), node(L1, Value, R1)) :-
  (Elt = Value ->
    L1 = L,
    R1 = R
  ;Elt < Value -> 
    R1 = R,
    add(Elt, L, L1)
  ;Elt > Value ->
    L1 = L, 
    add(Elt, R, R1)  
  ).

count_nodes(empty, 0).
count_nodes(node(L, _Value, R), Count) :- 
  count_nodes(L, LCount),
  count_nodes(R, RCount), 
  Count is LCount + RCount + 1.

sum_nodes(empty, 0.0).
sum_nodes(node(L, Value, R), Sum) :- 
  sum_nodes(L, LSum),
  sum_nodes(R, RSum),
  Sum is LSum + RSum + Value.

flatten(empty, []).
flatten(node(L, Value, R), Values) :- 
  flatten(L, LValues),
  flatten(R, RValues),
  append(LValues, RValues, LeafValues),
  append(LeafValues, [Value], Values).

average_value(empty, 0.0).
average_value(node(L, Value, R), Average) :-
  get_sum_count(node(L, Value, R), (Sum, Count)),
  Average is Sum / Count.

get_sum_count(empty, (0.0, 0.0)).
get_sum_count(node(L, Value, R), (Sum, Count)) :-
  get_sum_count(L, (LSum, LCount)),
  get_sum_count(R, (RSum, RCount)),
  Sum is LSum + RSum + Value,
  Count is LCount + RCount + 1.

average_value1(empty, 0.0). 
average_value1(node(L, Value, R), Average) :- 
  flatten(node(L, Value, R), Nodes),
  sum_list(Nodes, Sum),
  length(Nodes, Length),
  Average is Sum / Length.

average_value2(empty, 0.0).
average_value2(node(L, Value, R), Average) :-
  sum_nodes(node(L, Value, R), Sum),
  count_nodes(node(L, Value, R), Count),
  Average is Sum / Count.

same_shape(empty, empty).
same_shape(node(L, _, R), node(L1, _, R1)) :- 
  same_shape(L, L1),
  same_shape(R, R1). 
